fix(Email): pull seen/all mails via imap and decode subject

pstuhlmueller opened the pull request:
This PR holds four fixes in the context of fetching emails. Receiving emails has been refactored from v13 to v14, and by doing so, some issues have been introduced.

1. incorrect apostrophes for folder name: in docType "IMAP Folder" we store the folder names without apostrophes (eg. just "INBOX". By using an f-String within email_accoun.py we call the nested method with "'INBOX'", this will cause issues when fetching Mails via IMAP, as the UIDs will never be matched.
2. email_sync_option "ALL" did not work correctly (implicit set to "UNSEEN" recently), closes https://github.com/frappe/frappe/issues/18168 and closes https://github.com/frappe/frappe/issues/19059 
3. casting a string to int: the UIDs are integers but have been inconsistent processed as a string as well as int before, causing issues when doing comparisons. 
4. when there is an issue with recognizing the encoding of an email subject, this causes errors and breaks the import of new emails for an email account. This fix closes https://github.com/frappe/frappe/issues/20808 and closes https://github.com/frappe/frappe/issues/20714 . Furthermore it improves the idea of pending PR https://github.com/frappe/frappe/pull/20716

pstuhlmueller committed:
(fix) pull unseen mails via imap and decode subject

While refactoring the email logic from v13 to v14 a few things have been missed. Since then, only "unseen" mails could be pulled, see https://github.com/frappe/frappe/issues/19059 and https://github.com/frappe/frappe/issues/18168.

Furthermore decoding the subject was not working and sometimes only the first chunk has been encoded, see https://github.com/frappe/frappe/issues/20808 and https://github.com/frappe/frappe/pull/20716

pstuhlmueller committed:
(fix) helper method for proper decoding a sequence

Introduced new helpder method for decoding "subject" from emails, see https://github.com/frappe/frappe/issues/20808 and https://github.com/frappe/frappe/pull/20716

pstuhlmueller committed:
(fix) incorrect apostrophes for folder name

A folder setup as INBOX will be added hier into a string as '"INBOX"' instead of 'INBOX'. This will cause nested issues while updating the UIDs within check_imap_uidvalidity for docType "IMAP Folder" within receive.py.

pstuhlmueller committed:
add import for chardet

pstuhlmueller commented on frappe/email/receive.py:
```diff
@@ -215,23 +215,17 @@ def check_imap_uidvalidity(self, folder):
 				Communication.communication_medium == "Email"
 			).where(Communication.email_account == self.settings.email_account).run()
 
-			if self.settings.use_imap:
```
I have removed that block, as the whole method is dedicated to fetching mails via IMAP. I guess this has been missed, when refactoring the logic from v13 to v14 recently.

phot0n replied:
you're awesome - thanks for this - i've been meaning to remove this

pstuhlmueller committed:
(fix) removed irrelevant db update

UIDs in the context of IMAP are stored within "IMAP Folder". This is deprecated and (maybe) should have been noticed during refactoring.

phot0n commented on frappe/__init__.py:
```diff
@@ -2367,6 +2368,30 @@ def safe_decode(param, encoding="utf-8"):
 	return param
 
 
+def decode_sequence(encoded_sequence) -> str:
```
can this be moved to `frappe.utils.data.py`? this will probably not be a frequently used function so it's better not to add it to frappe namespace

pstuhlmueller replied:
Thats a good point. I will move that helper method.

phot0n commented on frappe/email/receive.py:
```diff
@@ -215,23 +214,17 @@ def check_imap_uidvalidity(self, folder):
+				self.settings.email_sync_rule = "ALL"
 				return
 
 			sync_count = 100 if uid_validity else int(self.settings.initial_sync_count)
```
let's remove the initial_sync_count as well? practically useless - we only slice 100 emails from the "list" of emails (currently)

phot0n replied:
Rather than inital sync count, lets give the control to user? Where they can control how many emails get pulled into the system at once? On the doctype itself?

Though we might need to configure the job's timeout then - lets just remove this for now i guess

will need to look at this a bit more closely - ignore for now

phot0n commented on frappe/email/receive.py:
```diff
@@ -200,13 +200,12 @@ def get_new_mails(self, folder):
 	def check_imap_uidvalidity(self, folder):
 		# compare the UIDVALIDITY of email account and imap server
-		uid_validity = self.settings.uid_validity
+		uid_validity = int(self.settings.uid_validity) if self.settings.uid_validity else 0
```
```suggestion
		uid_validity = cint(self.settings.uid_validity)
```

phot0n commented on frappe/email/receive.py:
```diff
@@ -200,13 +200,12 @@ def get_new_mails(self, folder):
 
 		response, message = self.imap.status(folder, "(UIDVALIDITY UIDNEXT)")
-		current_uid_validity = self.parse_imap_response("UIDVALIDITY", message[0]) or 0
+		current_uid_validity = int(self.parse_imap_response("UIDVALIDITY", message[0])) or 0
```
```suggestion
		current_uid_validity = cint(self.parse_imap_response("UIDVALIDITY", message[0]))
```

phot0n reviewed (commented):
I've yet to test this but things do seem to look promising - thanks for this

pstuhlmueller committed:
replace int with cint

Co-authored-by: Ritwik Puri <ritwikpuri5678@gmail.com>

pstuhlmueller committed:
replace int with cint

Co-authored-by: Ritwik Puri <ritwikpuri5678@gmail.com>

pstuhlmueller committed:
move helper method to utils.data

As discussed in PR #20860, we've decided to move this helper, as it is more seldom in use and should therefore not in the frappe namespace

pstuhlmueller committed:
(fix) helper method for proper decoding a sequence

Introduced new helpder method for decoding "subject" from emails, see #20808 and #20716

pstuhlmueller committed:
replace import for helper decode_sequence

pstuhlmueller committed:
(fix) added missing import

phot0n commented on frappe/utils/data.py:
```diff
@@ -2238,3 +2240,28 @@ def add_trackers_to_url(url: str, source: str, campaign: str, medium: str = "ema
+	Returns:
+		str: The decoded and concatenated sequence string.
+	"""
+	from frappe import safe_decode
```
this is not really needed? - we can just use `frappe.safe_decode` :)

phot0n commented on frappe/utils/data.py:
```diff
@@ -2238,3 +2240,28 @@ def add_trackers_to_url(url: str, source: str, campaign: str, medium: str = "ema
+	decoded_string = ""
+	for chunk, encoding in encoded_sequence:
+		if encoding is not None:
+			decoded_chunk = safe_decode(param=chunk, encoding=encoding)
```
i think in the [other pr](https://github.com/frappe/frappe/pull/20716), the encoding came out to be `unknown-8bit` , hence they used `chardet.detect` here?

phot0n commented on frappe/utils/data.py:
```diff
@@ -2238,3 +2240,28 @@ def add_trackers_to_url(url: str, source: str, campaign: str, medium: str = "ema
+		if encoding is not None:
+			decoded_chunk = safe_decode(param=chunk, encoding=encoding)
+		else:
+			if not isinstance(chunk, str):
+				detected_encoding = chardet.detect(chunk)["encoding"]
+				decoded_chunk = safe_decode(param=chunk, encoding=detected_encoding)
+			else:
+				decoded_chunk = safe_decode(param=chunk)
```
this branching is unnecessary? since there is no encoding present?

in decode_header's definition: https://github.com/python/cpython/blob/main/Lib/email/header.py#L62-L65

#

Also this whole helper feels very email specific - i initially thought it was generic - lets move it to 'email/utils.py'?

pstuhlmueller committed:
refactoring: undo changes

implementation moved to email/utils.py

pstuhlmueller committed:
(fix) helper for decoding email subjects

pstuhlmueller committed:
change import to latest implementation

pstuhlmueller committed:
(fix) import of helper

pstuhlmueller committed:
improved helper method

pstuhlmueller committed:
(fix): initial sync based in initial_sync_count

In the case of initial_sync/ reindexing, we have unexpected behavior, as we were fetching too many messages, as email_list (holding uids) is sorted asc, where the smallest uid = the oldest email. Therefore, when syncing ALL emails, we (a) have to reverse the order and (b) check for the initial_sync_count.

pstuhlmueller commented:
Hi @phot0n ,
I have updated the PR and added an improvement to get_messages, as there were issues when doing an initial_sync (see also https://discuss.frappe.io/t/email-initial-sync-count-ignored/99888). The reason was that we (also with the previous implementation with static 50 messages fetches, instead of the 100 after your refactoring) have to reverse the uid list to get the latest (newest) emails and then limit it to the initial sync_count (100 or 250 based on the settings).

pstuhlmueller committed:
refactoring and fix for initial sync

pstuhlmueller commented:
sorry for the additional commit, I had a merge conflict on my repo, before, and had to integrate my latest refactoring, again..

IgorA100 commented:
I tried applying changes to version:
_ERPNext: v14.24.0 (version-14)
Frappe Framework: v14.36.0 (version-14)
Frappe HR: v15.0.0-dev (develop)_
The initial sync does not work (email_sync_option = ALL).
When executing the "check_imap_uidvalidity" method, the variables are:
uid_validity=current_uid_validity=uidnext = 314493284

In DB table  "tabIMAP Folder":
uidnext  = 555902
uid_validity = 314493284

In DB table  "tabEmail Account":
uidnext  = 0
uidvalidity = ''

pstuhlmueller commented:
> I tried applying changes to version: _ERPNext: v14.24.0 (version-14) Frappe Framework: v14.36.0 (version-14) Frappe HR: v15.0.0-dev (develop)_ The initial sync does not work (email_sync_option = ALL). When executing the "check_imap_uidvalidity" method, the variables are: uid_validity=current_uid_validity=uidnext = 314493284
> 
> In DB table "tabIMAP Folder": uidnext = 555902 uid_validity = 314493284
> 
> In DB table "tabEmail Account": uidnext = 0 uidvalidity = ''

Regarding Email Account table: That's ok, as that information/table is only relevant for POP3-Sync, not for IMAP and won't be updated with IMAP sync any longer. 
There was also recently a patch that would have moved the (deprecated) info for IMAP folders from "Email Account" into "IMAP_Folder": https://github.com/frappe/frappe/blob/develop/frappe/patches/v14_0/copy_mail_data.py
That was important, as you could sync multiple imap_folders, while this wasn't possible before with POP3 sync, as the uid-info was only persistent within "Email Account".

ossplus commented:
> sorry for the additional commit, I had a merge conflict on my repo, before, and had to integrate my latest refactoring, again..

Is it ready for V14 production?

IgorA100 commented:
> There was also recently a patch that would have moved the (deprecated) info for IMAP folders from "Email Account" into "IMAP_Folder": https://github.com/frappe/frappe/blob/develop/frappe/patches/v14_0/copy_mail_data.py

1. I installed clean ERPNext: v14.24.0 & Frappe Framework: v14.36.0
2. I applied your fixes to version 14.
3. Only after that I added an email account! But synchronization did not start!
4. I tried bench --site mysite migrate & bench restart but that didn't work.
Probably the system, when executing the "check_imap_uidvalidity" method, does NOT correctly receive "uidnext = 314493284", although in the table "tabIMAP Folder" uidnext = 555902 !
Contents of SELF at the time "check_imap_uidvalidity" was executed:
{'settings': {'email_account_name': 'xxx', 'email_account': 'xxx', 'host': 'xxx', 'use_ssl': 0, 'use_starttls': 0, 'username': 'xxx@xxx ', 'use_imap': 1, 'email_sync_rule': 'UID 1:501', 'incoming_port': 143, 'initial_sync_count': '500', 'use_oauth': False, 'access_token': None, 'password': ' xxx', 'uid_validity': '314493284'}, 'imap': <frappe.email.receive.Timed_IMAP4 object at 0x7fbce0917ac0>, 'latest_messages': [], 'seen_status': {}, 'uid_reindexed': False}

IgorA100 commented:
@pstuhlmueller 
What can you say about the problem in version 14?

pstuhlmueller committed:
fix(decoding): check for type string

There might be concatenation issues in edge cases after decoding. I have added a check to ensure we only concatenate strings (and not concatenable bytes).

pstuhlmueller commented:
> > There was also recently a patch that would have moved the (deprecated) info for IMAP folders from "Email Account" into "IMAP_Folder": https://github.com/frappe/frappe/blob/develop/frappe/patches/v14_0/copy_mail_data.py
> 
> 1. I installed clean ERPNext: v14.24.0 & Frappe Framework: v14.36.0
> 2. I applied your fixes to version 14.
> 3. Only after that I added an email account! But synchronization did not start!
> 4. I tried bench --site mysite migrate & bench restart but that didn't work.
>    Probably the system, when executing the "check_imap_uidvalidity" method, does NOT correctly receive "uidnext = 314493284", although in the table "tabIMAP Folder" uidnext = 555902 !
>    Contents of SELF at the time "check_imap_uidvalidity" was executed:
>    {'settings': {'email_account_name': 'xxx', 'email_account': 'xxx', 'host': 'xxx', 'use_ssl': 0, 'use_starttls': 0, 'username': 'xxx@xxx ', 'use_imap': 1, 'email_sync_rule': 'UID 1:501', 'incoming_port': 143, 'initial_sync_count': '500', 'use_oauth': False, 'access_token': None, 'password': ' xxx', 'uid_validity': '314493284'}, 'imap': <frappe.email.receive.Timed_IMAP4 object at 0x7fbce0917ac0>, 'latest_messages': [], 'seen_status': {}, 'uid_reindexed': False}

What has worked for us in multiple dev environments as well as three production environments is:
1) Apply the changes, perform bench restart
2) For the relevant Email Account: make sure, no Communication documents existed before OR at least make sure, they have not a UID set
3) Setup the EMail Account
3.1) Optionally check "Email Account" table in the database as well as "IMAP Folder". You will see empty/null values for the UIDs
3.2) If you are using OAuth: Don't tick "Enable Incoming" in the first place. After creation, let the user authorize via OAuth Flow. Afterwards, tick "enable incoming" and set sync-option to "ALL"
4) Check the RQ Job Log and see, if the sync. job has been executed. Also check Error Log List, if there was an error while executing the sync.


Have you checked in general, if the job was executed? Have there been Communication documents for that Email Account on DB?

IgorA100 commented:
> Have you checked in general, if the job was executed? 

Yes! 
If the tasks were not executed, then how would I be able to get debugging information in the form of receiving an INCORRECT "uidnext" in the ""check_imap_uidvalidity"" method
uid_validity=current_uid_validity=uidnext = 314493284
uidnext should be 555902, not 314493284
> Have there been Communication documents for that Email Account on DB?

No.

I have created a new user.
I have added an email address
I enabled the inbox for the email address. Synchronize ALL messages.
Doesn't work in version 14!

IgorA100 commented:
It really works.
it was necessary to completely replace the "frappe/email/receive.py" file in version 14 with the develop version.
I keep testing.

tobehn commented:
Will this be merged so we finally do not mark every email as read by the initial synchronisation?

IgorA100 commented:
If there is an approval of this PR, then IMHO - this fix should be applied to version 14 as well
My testing on version 14 found no issues.

pstuhlmueller commented:
From my side, there are also no additional improvements pending. We are using this patched version now for a couple of weeks in prod without issues.

this-gavagai commented:
I can also attest that this PR is working for us in v14. If there is anything we can do to help get this merged, please let us know. I appreciate the need for code review, but incoming email has been broken in v14 since the beginning.

IgorA100 commented on frappe/email/utils.py:
```diff
@@ -16,3 +18,25 @@ def get_port(doc):
+	decoded_string = ""
+	for chunk, encoding in encoded_sequence:
+		decoded_chunk = ""
+		if not isinstance(chunk, str):
```
Maybe it is necessary to reduce the redundant code to this?
```
		if not isinstance(chunk, str):
			detected_encoding = encoding if encoding is not None and encoding != 'unknown-8bit' else chardet.detect(chunk)["encoding"]
			decoded_string += chunk.decode(detected_encoding, 'ignore')
		else:
			decoded_string += chunk
```
1. If  type(chunk) = str, then it has already been decoded and it is not worth trying to decode it again.
2. Use condition:
```
		if isinstance(decoded_chunk, str):
			decoded_string += decoded_chunk
```
not the best option, because in case of a decoding error, we will completely lose the entire chunk. I think it would be better to use **chunk.decode(detected_encoding, 'ignore')** instead of **safe_decode(param=chunk, encoding=detected_encoding)** so most of the chunk will be decoded correctly.

IgorA100 replied:
???

IgorA100 replied:
Am I wrong about something?
